### DISCUSSION TITLE
Am I doing something wrong in trying to receive TCP message over network?

### DISCUSSION THREAD

**Discussion by chengguizi** [eclipse-ecal/ecal#975] [Q&A] [answered]
I am current trying to use eCAL as middleware to stream image topics over network, between two computers (Ethernet cable).

As instructed by the documentation here https://eclipse-ecal.github.io/ecal/advanced/layers/tcp.html?highlight=tcp , I have setup 
```
use_tcp                   = 2
use_udp_mc                = 0
```

I am able creating a publisher on machine A, and machine A could receive message properlly on A itself. However something strange is on Machine B:
1. ecal Monitor could see the message on the other machine A (with stats)
2. However, there is no reception of actual message

stats visible
![image](https://user-images.githubusercontent.com/7508112/219622009-13a98f9b-fab9-4ab6-929c-58f818e9cdf5.png)

but no message
![image](https://user-images.githubusercontent.com/7508112/219622709-70c0d70f-0224-4187-9d7e-ef8cb104f651.png)

Am I doing anything wrong / missing?

**Comment by chengguizi** [comment]
I also tried the stock sample.

On machine A
```
ecal_sample_person_snd_tcp
CTCPReaderLayer - TCPPubSub (Error) -Publisher ?: Error while waiting for subsriber: Operation aborted.
person id    : 1
person name  : Max
person stype : 0
person email : max@mail.net
dog.name     : Brandy
house.rooms  : 4

person id    : 2
person name  : Max
person stype : 0
person email : max@mail.net
dog.name     : Brandy
house.rooms  : 4

person id    : 3
person name  : Max
person stype : 0
```

Also on Machine A:
```
ecal_sample_person_rec
------------------------------------------
 HEAD 
------------------------------------------
topic name   : person
topic time   : 1676615959532414
topic clock  : 3
------------------------------------------
 CONTENT 
------------------------------------------
person id    : 3
person name  : Max
person stype : 0
person email : max@mail.net
dog.name     : Brandy
house.rooms  : 4
```

But on Machine B, `ecal_sample_person_rec` is not able to obtain any message

**Comment by chengguizi** [comment]
I can catch some warning on eCAL Mon
```
2023-02-17 11:28:45.905	Warning	eCALMon	CTCPReaderLayer - TCPPubSub (Warning) -SubscriberSession ?->NUC12WSHi5:38843: Failed to resolve address: Host not found (non-authoritative), try again later

```

Then I dig further to this link https://eclipse-ecal.github.io/ecal/getting_started/services.html?highlight=hostname#hostname-resolution

I have to add the hostname to /etc/hosts. And now things seem to work.

Do I really have to manually configure `/etc/hosts` for the tcp pub/sub to work? That would be quite troublesome and error-prone..

**Comment by FlorianReimold** [comment, marked as answer]
Hi @chengguizi,

Glad to see you already figured out a solution.

So yes: You have to add a Hostname <-> IP mapping in your `/etc/hosts` file. Or more precisely: you have to provide that mapping to your operating system in any manner possible. If you were on 2 Windows machines, that mapping would happen automatically by Windows. But Linux just doesn't do that.
https://eclipse-ecal.github.io/ecal/getting_started/services.html#hostname-resolution

eCAL uses Hostnames to address remote machines. It does that for UDP and TCP. But for UDP, eCAL uses multicast and therefore doesn't need to know the IP address of the remote host. TCP on the other hand is a point-to-point connection, which requires the IP Adress to be resolvable.

Kind Regards
Florian
